### ISSUE TITLE
LSF can return an inconsistent jobs status  

### ISSUE THREAD

**Issue by pditommaso** [nextflow-io/nextflow#927]
LSF `bjobs` command can return an invalid task status in a non-deterministic manner causing the workflow execution to error by a false positive task failure. 

More in detail NF periodically run bjobs to fetch tasks status, eg.

```
Nov-13 12:41:59.838 [Task monitor] TRACE n.executor.AbstractGridExecutor - [LSF] getting queue status > cmd: bjobs -w
Nov-13 12:41:59.861 [Task monitor] TRACE n.executor.AbstractGridExecutor - [LSF] queue status > cmd exit: 0
JOBID   USER    STAT  QUEUE      FROM_HOST   EXEC_HOST   JOB_NAME   SUBMIT_TIME
5132676 pluskal RUN   normal     tak         it-c01b04   /lab/solexa_weng/testtube/transcriptome/check_strandedness.sh SRR2006304_1.fastq.gz SRR2006307_1_fixed.fastq.gz Nov 13 00:41
5137404 pluskal RUN   normal     it-c03b03   it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy nf-trinityInchwormChrysalis Nov 13 12:18
5137399 pluskal RUN   normal     tak         it-c03b03   NXF_VER=18.11.0-edge nextflow -trace nextflow.executor run assemble.nf Nov 13 12:17
```

Some minutes later bjobs returns an empty result:

```
Nov-13 12:42:59.840 [Task monitor] TRACE n.executor.AbstractGridExecutor - [LSF] getting queue status > cmd: bjobs -w
Nov-13 12:42:59.861 [Task monitor] TRACE n.executor.AbstractGridExecutor - [LSF] queue status > cmd exit: 0

```

However the jobs it's still running, indeed later in the log, it appears again: 

```
Nov-13 12:44:59.853 [Task monitor] TRACE n.executor.AbstractGridExecutor - [LSF] getting queue status > cmd: bjobs -w
Nov-13 12:44:59.870 [Task monitor] TRACE n.executor.AbstractGridExecutor - [LSF] queue status > cmd exit: 0
JOBID   USER    STAT  QUEUE      FROM_HOST   EXEC_HOST   JOB_NAME   SUBMIT_TIME
5132676 pluskal RUN   normal     tak         it-c01b04   /lab/solexa_weng/testtube/transcriptome/check_strandedness.sh SRR2006304_1.fastq.gz SRR2006307_1_fixed.fastq.gz Nov 13 00:41
5137404 pluskal RUN   normal     it-c03b03   it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy:it-bigboy nf-trinityInchwormChrysalis Nov 13 12:18
5137399 pluskal RUN   normal     tak         it-c03b03   NXF_VER=18.11.0-edge nextflow -trace nextflow.executor run assemble.nf Nov 13 12:17

```

However the empty result makes NF to mark that jobs as into an inconsistent status. It set a timeout make it to fail fast when as soon the job exit (successfully) later, because the `.exitcode` is not created (yet apparently): 

```
Content of workDir: /lab/weng_scratch/Tomas/nextflow_assemble/work/ba/cbcdeac0e24debb012453375935c56
> total 204
> drwxrwsr-x 3 pluskal wenglab  306 Nov 13 12:18 .
> drwxrwsr-x 3 pluskal wenglab   48 Nov 13 12:18 ..
> -rw-rw-r-- 1 pluskal wenglab    0 Nov 13 12:18 .command.begin
> -rw-rw-r-- 1 pluskal wenglab    0 Nov 13 12:18 .command.err
> -rw-rw-r-- 1 pluskal wenglab 7626 Nov 13 13:04 .command.log
> -rw-rw-r-- 1 pluskal wenglab 7626 Nov 13 13:04 .command.out
> -rw-rw-r-- 1 pluskal wenglab 2921 Nov 13 12:18 .command.run
> -rw-rw-r-- 1 pluskal wenglab  247 Nov 13 12:18 .command.sh
> -rw-rw-r-- 1 pluskal wenglab 3500 Nov 13 12:18 .command.stub
> -rw-rw-r-- 1 pluskal wenglab  210 Nov 13 13:08 .command.trace
> lrwxrwxrwx 1 pluskal wenglab   53 Nov 13 12:18 samples.txt -> /lab/weng_scratch/Tomas/nextflow_assemble/samples.txt
> drwxrwsr-x 3 pluskal wenglab  705 Nov 13 13:04 trinity_out_dir
```  

[nextflow_log.txt](https://github.com/nextflow-io/nextflow/files/2580403/nextflow_log.txt)

**Comment by pditommaso** [comment]
@tomas-pluskal  You may want to give another try with this version `18.12.0-SNAPSHOT`.

**Comment by tomas-pluskal** [comment]
I tried to run it with 18.12.0-SNAPSHOT. The first run crashed with a NullPointerException.
Then I ran it again and it finished fine. The logs are attached below.

[nextflow_log_1.txt](https://github.com/nextflow-io/nextflow/files/2581686/nextflow_log_1.txt)
[nextflow_log_2.txt](https://github.com/nextflow-io/nextflow/files/2581687/nextflow_log_2.txt)

**Comment by pditommaso** [comment]
The error stack trace it's completely unrelated the bjobs issue. 

```
Nov-14 09:50:10.362 [Actor Thread 26] ERROR n.extension.DataflowExtensions - @unknown
java.lang.NullPointerException: null
	at org.iq80.leveldb.impl.SnapshotImpl.<init>(SnapshotImpl.java:35)
	at org.iq80.leveldb.impl.DbImpl.getSnapshot(DbImpl.java:805)
	at org.iq80.leveldb.impl.DbImpl.get(DbImpl.java:588)
	at org.iq80.leveldb.impl.DbImpl.get(DbImpl.java:577)
	at nextflow.sort.LevelDbSort.get(LevelDbSort.java:80)
	at nextflow.sort.BigSort$SliceComparator.get(BigSort.java:424)
	at nextflow.sort.BigSort$SliceComparator.compare(BigSort.java:446)
	at java.util.TimSort.countRunAndMakeAscending(TimSort.java:356)
	at java.util.TimSort.sort(TimSort.java:234)
	at java.util.Arrays.sort(Arrays.java:1438)
	at nextflow.sort.BigSort.sliceSort(BigSort.java:280)
	at nextflow.sort.BigSort.sort(BigSort.java:235)
	at nextflow.file.SortFileCollector.makeHash(SortFileCollector.groovy:247)
	at nextflow.file.FileCollector.saveTo(FileCollector.groovy:238)
	at nextflow.file.FileCollector$saveTo$0.call(Unknown Source)
	at nextflow.extension.CollectFileOp.emitItems(CollectFileOp.groovy:192)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:104)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:326)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1217)
	at groovy.lang.MetaClassImpl.invokeMethodClosure(MetaClassImpl.java:1050)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1092)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at groovy.lang.Closure.call(Closure.java:411)
	at groovy.lang.Closure.call(Closure.java:427)
	at groovy.lang.Closure$call.call(Unknown Source)
	at nextflow.extension.DataflowHelper$2.afterStop(DataflowHelper.groovy:202)
	at groovyx.gpars.dataflow.operator.DataflowProcessor.fireAfterStop(DataflowProcessor.java:324)
	at groovyx.gpars.dataflow.operator.DataflowProcessorActor.afterStop(DataflowProcessorActor.java:59)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:104)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:326)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1217)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePojoMethod(InvokerHelper.java:999)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:990)
	at groovyx.gpars.actor.Actor.callDynamic(Actor.java:369)
	at groovyx.gpars.actor.Actor.handleTermination(Actor.java:320)
	at groovyx.gpars.actor.AbstractLoopingActor$1.handleMessage(AbstractLoopingActor.java:86)
	at groovyx.gpars.util.AsyncMessagingCore.run(AsyncMessagingCore.java:132)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Actually I've never seen before. It seems related to use of a `collectFile` operator. Maybe there's a null value?

**Comment by tomas-pluskal** [comment]
Ok, perhaps the bjobs issue is solved then, and this is a separate issue?

**Comment by pditommaso** [comment]
If you are able to replicate the error consistently, yes please open a separate issue.

**Comment by tomas-pluskal** [comment]
Like I said, the second run finished successfully, so this issue seems to be non-deterministic. I will try a few more runs and see what happens.

**Comment by tomas-pluskal** [comment]
I ran the whole workflow about 5 times and have not seen the NullPointerException again. Not sure where it came from originally.

**Comment by tomas-pluskal** [comment]
Hi,
I am afraid this issue is has not been fully fixed. I still get jobs failing with
```
ERROR ~ Error executing process > 'trinityInchwormChrysalis'

Caused by:
 Process `trinityInchwormChrysalis` terminated for an unknown reason -- Likely it has been terminated by the external system
```

I have attached the log file of one such run, executed with ```NXF_VER=18.12.0-SNAPSHOT nextflow -trace nextflow.executor run transXpress.nf -resume```

[nextflow.log.txt](https://github.com/nextflow-io/nextflow/files/2596548/nextflow.log.txt)

According to the log, nextflow determined that job 5182892 had ended. 

However, the job is still running. Attached below is the full output of ```bjobs -w``` ~30 min after the nextflow run failed.

[bjobs.txt](https://github.com/nextflow-io/nextflow/files/2596551/bjobs.txt)

Could this be a timing issue? Does nextflow have a time limit for bjobs to respond?

**Comment by tomas-pluskal** [comment]
Note that it is possible to run bjobs for each job separately by adding the job ID on the command line. Using the -a parameter also allows to include recently finished jobs.

```
$ bjobs -a 5182892
JOBID   USER    STAT  QUEUE      FROM_HOST   EXEC_HOST   JOB_NAME   SUBMIT_TIME
5182892 pluskal RUN   normal     it-c01b14   it-bigboy   *Chrysalis Nov 19 12:57
                                             it-bigboy
                                             it-bigboy
                                             it-bigboy
                                             it-bigboy
                                             it-bigboy
                                             it-bigboy
                                             it-bigboy
                                             it-bigboy
                                             it-bigboy
```

Example of a finished job:
```
$ bjobs -a 5182966
JOBID   USER    STAT  QUEUE      FROM_HOST   EXEC_HOST   JOB_NAME   SUBMIT_TIME
5182966 pluskal DONE  normal     it-c01b14   it-c05b10   *el.208.sh Nov 19 13:22
                                             it-c05b10
```

Perhaps this could be a better strategy for NF to determine the status of individual jobs?

**Comment by tomas-pluskal** [comment]
@pditommaso Can we reopen this issue?

**Comment by pditommaso** [comment]
Yes

**Comment by tomas-pluskal** [comment]
So based on the log files it seems pretty clear that running ```bjobs -w``` can occasionally be unreliable, with some jobs simply being unreported. I think nextflow should be instead running ```bjobs -a -w``` (or ```bjobs -a -w [jobid]```), which will include the DONE or EXIT status for recently finished jobs. The most important thing is that if nextflow fails to determine the status of a job, it should assume the job is still running, and simply try again later. 

What do you think about this proposal?

**Comment by pditommaso** [comment]
It's quite clear that before the execution is terminated the job `5182892` is not returned by `bjobs`: 

```
Nov-19 13:25:59.624 [Task monitor] TRACE n.executor.AbstractGridExecutor - [LSF] queue status > cmd exit: 0
JOBID   USER    STAT  QUEUE      FROM_HOST   EXEC_HOST   JOB_NAME   SUBMIT_TIME
5183028 pluskal RUN   normal     it-c01b14   it-c01b11:it-c01b11 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.162.sh Nov 19 13:22
5182934 pluskal RUN   normal     it-c01b14   it-c01b04:it-c01b04 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.166.sh Nov 19 13:22
5183047 pluskal RUN   normal     it-c01b14   it-c06b01:it-c06b01 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.181.sh Nov 19 13:22
5183066 pluskal RUN   normal     it-c01b14   it-c06b01:it-c06b01 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.156.sh Nov 19 13:22
5183084 pluskal RUN   normal     it-c01b14   it-c06b01:it-c06b01 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.182.sh Nov 19 13:22
5183104 pluskal RUN   normal     it-c01b14   it-c06b01:it-c06b01 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.157.sh Nov 19 13:22
5183022 pluskal RUN   normal     it-c01b14   it-c06b12:it-c06b12 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.153.sh Nov 19 13:22
5183069 pluskal RUN   normal     it-c01b14   it-c06b12:it-c06b12 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.158.sh Nov 19 13:22
5183087 pluskal RUN   normal     it-c01b14   it-c06b12:it-c06b12 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.184.sh Nov 19 13:22
5183092 pluskal RUN   normal     it-c01b14   it-c06b10:it-c06b10 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.171.sh Nov 19 13:22
5182984 pluskal RUN   normal     it-c01b14   it-c06b09:it-c06b09 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.170.sh Nov 19 13:22
5183098 pluskal RUN   normal     it-c01b14   it-c06b09:it-c06b09 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.167.sh Nov 19 13:22
5182986 pluskal RUN   normal     it-c01b14   it-c06b05:it-c06b05 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.175.sh Nov 19 13:22
5183056 pluskal RUN   normal     it-c01b14   it-c06b05:it-c06b05 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.189.sh Nov 19 13:22
5183075 pluskal RUN   normal     it-c01b14   it-c06b05:it-c06b05 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.168.sh Nov 19 13:22
5183099 pluskal RUN   normal     it-c01b14   it-c06b05:it-c06b05 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.207.sh Nov 19 13:22
5183051 pluskal RUN   normal     it-c01b14   it-c06b07:it-c06b07 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.195.sh Nov 19 13:22
5183085 pluskal RUN   normal     it-c01b14   it-c06b07:it-c06b07 /lab/solexa_weng/Seq_data/Projects/Tomas_Pluskal/20181112_Euphorbia_peplus_SRR2014663/.snakemake/tmp.a3hy3x2h/snakejob.pfam_parallel.183.sh Nov 19 13:22
```

Not sure what could be the cause, I suspect a problem with your LSF version. There are other users using LSF 9 and I've never seen this kind of problem. 

`bjobs -w [jobid]` is surely not an option because it would require to execute *n* `bjobs` invocation that can be overkilling with large workflows and many concurrent instances. 

I have no idea if the `-a` can help.

**Comment by tomas-pluskal** [comment]
I agree that this is clearly a bug in LSF. But it not realistic for me to demand from our IT staff to rebuild our cluster just because of this bug. I think nextflow has very naive expectations about its environment working properly. The truth is we don't live in a perfect software world, and nextflow should expect that any command may fail (including bjobs), and if the required answer could not be obtained, it should simply try again later or assume that the job of unknown status may still be running.

**Comment by pditommaso** [comment]
> I think nextflow has very naive expectations about its environment 

I don't agree at all, nextflow takes very seriously this kind of problems. Nevertheless, this does not mean it should be the sink of any broken software

**Comment by pditommaso** [comment]
Included in version `18.12.0-edge`.
